compiler/src/main/kotlin/io/github/aplcornell/viaduct/backends/cleartext/CleartextCircuitCodeGenerator.kt

```diff
@@ -112,6 +121,128 @@ class CleartextCircuitCodeGenerator(context: CodeGeneratorContext) : AbstractCod
+                builder.addStatement(
+                    "val %N = %N.%M()",
+                    tempName2,
+                    tempName1,
+                    MemberName(Committed.Companion::class.asClassName(), "commitment"),
+                )
+                receivingHosts.forEach {
+                    builder.addStatement("%L", context.send(CodeBlock.of("%N", tempName2), it))
+                }
+                CodeBlock.of("%N", tempName1)
+            }
+
+            in receivingHosts -> {
+                val tempName3 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %L",
+                    tempName3,
+                    context.receive((Commitment::class).asTypeName().parameterizedBy(argType), source.host),
+                )
+                CodeBlock.of("%N", tempName3)
+            }
+
+            else -> throw IllegalStateException()
+        }
+    }
+
+    private fun openCommitment(
+        source: Protocol,
+        target: Protocol,
+        argument: Argument,
+        builder: CodeBlock.Builder,
+    ): CodeBlock {
+        require(source is CommitmentProtocol)
+        if (target !is Cleartext) {
+            throw UnsupportedCommunicationException(source, target, argument.sourceLocation)
+        }
+        require(context.host in source.hosts + target.hosts)
+        if (source.hashHosts != target.hosts || source.cleartextHost in source.hashHosts) {
+            throw UnsupportedCommunicationException(source, target, argument.sourceLocation)
+        }
+
+        val argType = kotlinType(argument.type.shape, typeTranslator(argument.type.elementType.value))
+        val sendingHost = source.cleartextHost
+        val receivingHosts = target.hosts
+        return when (context.host) {
+            sendingHost -> {
+                receivingHosts.forEach {
+                    builder.addStatement("%L", context.send(argument.value, it))
+                }
+                CodeBlock.of("%L.value", argument.value)
+            }
+            in receivingHosts -> {
+                val tempName1 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %L",
+                    tempName1,
+                    context.receive((Committed::class).asTypeName().parameterizedBy(argType), source.cleartextHost),
+                )
+                val tempName2 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %L",
+                    tempName2,
+                    argument.value,
+                )
+                val tempName3 = context.newTemporary("CommitTemp")
+                builder.addStatement(
+                    "val %N = %N.%N(%N)",
+                    tempName3,
+                    tempName2,
+                    "open",
+                    tempName1,
+                )
+
+                val peers = receivingHosts.filter { it != context.host }
```

Comment:
If you assume `hashHosts = receivingHosts` (which you do), the openers don't need to check with each other. They are guaranteed to get the correct value due to the binding property of commitments.